scene/animation/animation_node_state_machine.cpp

```diff
@@ -1603,12 +1642,13 @@ Vector2 AnimationNodeStateMachine::get_graph_offset() const {
 AnimationNode::NodeTimeInfo AnimationNodeStateMachine::_process(const AnimationMixer::PlaybackInfo p_playback_info, bool p_test_only) {
 	Ref<AnimationNodeStateMachinePlayback> playback_new = get_parameter(playback);
 	ERR_FAIL_COND_V(playback_new.is_null(), AnimationNode::NodeTimeInfo());
-	playback_new->_set_grouped(state_machine_type == STATE_MACHINE_TYPE_GROUPED);
+	playback_new->_set_base_path(node_state.get_base_path());
+	playback_new->_set_grouped(this);
```

Comment:
As I commented above, this needs to be done on specific setters and on the changed signal of the animation tree, not in process.

There may not be much of a performance issue with doing the bool check in process each time.

Rather, it is a matter of consistency. As far as I know, there is no other class that implements bind in process that should be associated with a setter. In the extreme case, since they are not done immediately after the setter, any processing between the setter and process could cause problems if the states do not match.

Reply:
The problem with using the setter + signals is that at no point besides process does the state machine or playback have reference to the tree, and so there's no parent playback to reference. I thought this change was appropriate as the playback's state and signals only change/emit during process (ie, it's already possible for the sm and playback to have mismatched states between process calls because the playback is always updated during process) and the original concern was characterized as `to avoid doing this too often` rather than process being entirely off limits. With that being the case, I'll change the code to avoid signals entirely and cache the parent playback.

Reply:
I see, then it may be acceptable, but the following approach may be more preferable.

Reply:
I'll finish up a version which bubbles up events from grouped SMs via method calls rather than signals for comparison. There's also a small benchmark I'd like to do for a related part of the code. Also while working on the non-signal version I realized this doesn't disconnect/reconnect on tree changes, so I'll change that as well.

Reply:
The direct callback approach ended up being fairly clean. I experimented a bit with caching the tree + parent context on the playback instance but it was an extremely marginal improvement relative to the complexity that introduced.